### PR TITLE
core: fix bugs in service config error handling

### PR DESCRIPTION
**Pull request by creamsoup**
~1. Allow legacy way to pass service config via attributes~
2. fix AutoConfiguredLBFactory may populate ConfigOrError instead of
actual config.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory2.java [hunk 1]
```diff
@@ -167,7 +167,7 @@ Status tryHandleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
             ResolvedAddresses.newBuilder()
                 .setAddresses(resolvedSelection.serverList)
                 .setAttributes(attributes)
-                .setLoadBalancingPolicyConfig(lbConfig)
+                .setLoadBalancingPolicyConfig(lbConfig != null ? lbConfig.getConfig() : null)
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory2.java]: Looks like there isn't a case where `selection.config` is an error. We may just change `PolicySelection.config` from `ConfigOrError` to `Object`.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java [hunk 2]
```diff
@@ -1332,9 +1333,13 @@ public void run() {
           ConfigOrError configOrError = resolutionResult.getServiceConfig();
           ServiceConfigHolder validServiceConfig = null;
           Status serviceConfigError = null;
+          Map<String, ?> rawServiceConfig =
+              resolutionResult.getAttributes().get(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG);
+          if (configOrError == null && rawServiceConfig != null) {
+            // use service config json in deprecated attributes
+            configOrError = serviceConfigParser.parseServiceConfig(rawServiceConfig);
+          }
           if (configOrError != null) {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java]: I think the original code is correct.

For a LoadBalancer that's still on the old service config path, it won't override `LoadBalancerProvider.parseLoadBalancingPolicyConfig`, thus `configOrError` would be `LoadBalancerProvider.UNKNOWN_CONFIG`, which is non-null, thus `NAME_RESOLVER_SERVICE_CONFIG` will get populated. We shouldn't call `parseServiceConfig` again.

Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java]: Some 1st/3rd party name resolver might not use the `serviceConfig` field which can be ignored in current implementation even if the deprecated `Attribute` is set. 
`LoadBalancerProvider.UNKNOWN_CONFIG` is still a valid config from the name resolver point of view, thus those aren't consumed.

Comment (zhangkun83) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java]: Now I see this is for "old" `NameResolver`s that don't call `parseServiceConfig()` at all but emit the raw config. Granted the raw config attribute is defined in the internal `GrpcAttributes` thus no third-party NameResolvers are expected to do that. It's fine to call parser here if that makes the migration for first-party NameResolver easier. It's going to be removed along with `GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG` anyway so not a big deal.

File: core/src/test/java/io/grpc/internal/ServiceConfigErrorHandlingTest.java [hunk 3]
```diff
@@ -541,7 +571,17 @@ public NameResolver newNameResolver(final URI targetUri, NameResolver.Args args)
         return null;
       }
       assertEquals(DEFAULT_PORT, args.getDefaultPort());
-      FakeNameResolver resolver = new FakeNameResolver(args.getServiceConfigParser());
+      FakeNameResolver resolver;
+      if (enableServiceConfigParsing) {
```
Comment (zhangkun83) [inline core/src/test/java/io/grpc/internal/ServiceConfigErrorHandlingTest.java]: An "old" NameResolver that doesn't do service config parsing should call `listener.onAddresses()`, not `onResult()`. This makes sure the compatibility logic in `Listener2#onAddresses()` is exercised.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java (hunk 2 above)
Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java]: it turns out that internal migration was easy. Thus, as pointed out the attributes shouldn't be used out side of gRPC. removing the service config parsing in `ManagedChannelImpl2`.

**Pull request merged**